## best score 0.88

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import FunctionTransformer, LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.model_selection import GridSearchCV

In [2]:
train = pd.read_csv('Data/train_with_BMI.csv')
test = pd.read_csv('Data/test_with_BMI.csv')
val = pd.read_csv('Data/val_with_BMI.csv')

In [3]:
X_train = train.drop('NObeyesdad', axis=1)
y_train = train['NObeyesdad']
X_test = test.drop('NObeyesdad', axis=1)
y_test = test['NObeyesdad']
X_val = val.drop('NObeyesdad', axis=1)
y_val = val['NObeyesdad']

In [4]:
#some random parameters
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)
y_pred1 = rf.predict(X_val)
class_report1 = classification_report(y_val, y_pred1)
print("Classification Report:\n", class_report1)
conf_matrix1 = confusion_matrix(y_val, y_pred1)
print("\nConfusion Matrix:")
print(conf_matrix1)

Classification Report:
               precision    recall  f1-score   support

           0       0.94      0.89      0.92       525
           1       0.81      0.88      0.85       624
           2       0.86      0.82      0.84       604
           3       0.94      0.96      0.95       714
           4       1.00      1.00      1.00       837
           5       0.73      0.73      0.73       532
           6       0.74      0.72      0.73       523

    accuracy                           0.87      4359
   macro avg       0.86      0.86      0.86      4359
weighted avg       0.87      0.87      0.87      4359


Confusion Matrix:
[[468  52   0   0   0   3   2]
 [ 24 551   0   0   0  40   9]
 [  1   3 496  30   4  29  41]
 [  0   0  22 686   0   0   6]
 [  2   0   0   1 834   0   0]
 [  1  53  16   0   0 387  75]
 [  0  18  46  12   0  68 379]]


In [5]:
clf = RandomForestClassifier(random_state=42)

RandomForestClassifier( )

param_grid = {
    'n_estimators':  [ 300,400,450, 500],
    'criterion' : ['gini', 'entropy'],
    'max_depth': [20, 25, 30],
    'n_jobs':[-1], 
    'min_samples_split':[ 2, 3,4, 5, 6,7,],
    'max_features': ['auto', 'sqrt', 'log2']

}

grid_search = GridSearchCV(estimator=clf, param_grid=param_grid, cv=3, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)


best_params = grid_search.best_params_
best_score = grid_search.best_score_
cv_results = grid_search.cv_results_

print("Best Parameters:", best_params)
print("Best Score:", best_score)
print("Best model CV Results:", cv_results["mean_test_score"][grid_search.best_index_])

Fitting 3 folds for each of 432 candidates, totalling 1296 fits
[CV] END criterion=gini, max_depth=20, max_features=auto, min_samples_split=2, n_estimators=300, n_jobs=-1; total time=   0.0s
[CV] END criterion=gini, max_depth=20, max_features=auto, min_samples_split=2, n_estimators=450, n_jobs=-1; total time=   0.0s
[CV] END criterion=gini, max_depth=20, max_features=auto, min_samples_split=2, n_estimators=500, n_jobs=-1; total time=   0.0s
[CV] END criterion=gini, max_depth=20, max_features=auto, min_samples_split=2, n_estimators=500, n_jobs=-1; total time=   0.0s
[CV] END criterion=gini, max_depth=20, max_features=auto, min_samples_split=2, n_estimators=500, n_jobs=-1; total time=   0.0s
[CV] END criterion=gini, max_depth=20, max_features=auto, min_samples_split=3, n_estimators=300, n_jobs=-1; total time=   0.0s
[CV] END criterion=gini, max_depth=20, max_features=auto, min_samples_split=3, n_estimators=300, n_jobs=-1; total time=   0.0s
[CV] END criterion=gini, max_depth=20, max_feat

/opt/anaconda3/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:547: FitFailedWarning: 
432 fits failed out of a total of 1296.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
343 fits failed with the following error:
Traceback (most recent call last):
  File "/opt/anaconda3/lib/python3.11/site-packages/sklearn/model_selection/_validation.py", line 895, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/opt/anaconda3/lib/python3.11/site-packages/sklearn/base.py", line 1467, in wrapper
    estimator._validate_params()
  File "/opt/anaconda3/lib/python3.11/site-packages/sklearn/base.py", line 666, in _validate_params
    validate_parameter_constraints(
  File "/opt/anaconda3/lib/python3.11/site-

Best Parameters: {'criterion': 'entropy', 'max_depth': 20, 'max_features': 'sqrt', 'min_samples_split': 6, 'n_estimators': 300, 'n_jobs': -1}
Best Score: 0.8858915899230326
Best model CV Results: 0.8858915899230326


BEST:  
Best Parameters: {'criterion': 'entropy', 'max_depth': 20, 'max_features': 'sqrt', 'min_samples_split': 6, 'n_estimators': 300, 'n_jobs': -1}
Best Score: 0.8858915899230326
Best model CV Results: 0.8858915899230326

In [7]:
#best
best = RandomForestClassifier(criterion='entropy', max_depth=20,min_samples_split=6,n_estimators=300,n_jobs=-1 , random_state=42, max_features='sqrt')
best.fit(X_train, y_train)
y_pred1 = best.predict(X_val)
class_report1 = classification_report(y_val, y_pred1)
print("Classification Report:\n", class_report1)
conf_matrix1 = confusion_matrix(y_val, y_pred1)
print("\nConfusion Matrix:")
print(conf_matrix1)

Classification Report:
               precision    recall  f1-score   support

           0       0.94      0.90      0.92       525
           1       0.82      0.89      0.85       624
           2       0.87      0.83      0.85       604
           3       0.94      0.97      0.95       714
           4       1.00      1.00      1.00       837
           5       0.75      0.73      0.74       532
           6       0.76      0.75      0.75       523

    accuracy                           0.88      4359
   macro avg       0.87      0.86      0.87      4359
weighted avg       0.88      0.88      0.88      4359


Confusion Matrix:
[[470  50   0   0   0   3   2]
 [ 25 555   1   0   0  38   5]
 [  2   2 500  33   4  25  38]
 [  0   0  17 691   0   0   6]
 [  1   0   0   1 834   1   0]
 [  0  56  14   0   0 387  75]
 [  0  14  44   9   0  64 392]]
